rust/src/tests/unlink.rs

```diff
@@ -0,0 +1,136 @@
+use nix::{
+    errno::Errno,
+    sys::stat::{fstat, lstat},
+    unistd::unlink,
+};
+
+use crate::{
+    runner::context::{FileType, SerializedTestContext, TestContext},
+    tests::{assert_ctime_changed, assert_ctime_unchanged},
+    utils::link,
+};
+
+use super::assert_mtime_changed;
+
+crate::test_case! {
+    /// unlink removes regular, block and char files, symbolic links, fifos and sockets
+    // unlink/00.t
+    remove_type => [Regular, Block, Char, Symlink(None), Fifo, Socket]
+}
+fn remove_type(ctx: &mut TestContext, ft: FileType) {
+    let path = ctx.create(ft).unwrap();
+
+    assert!(unlink(&path).is_ok());
+    assert_eq!(lstat(&path).unwrap_err(), Errno::ENOENT);
+}
+
+crate::test_case! {
+    /// successful unlink(2) updates ctime.
+    // unlink/00.t
+    update_ctime_success => [Regular, Block, Char, Fifo, Socket]
+}
+fn update_ctime_success(ctx: &mut TestContext, ft: FileType) {
+    let path = ctx.create(ft).unwrap();
+
+    let link_path = ctx.base_path().join("link");
+    link(&path, &link_path).unwrap();
+
+    assert_ctime_changed(ctx, &link_path, || {
+        assert!(unlink(&path).is_ok());
+    });
+}
+
+// TODO: why it isn't in the original test suite?
+// crate::test_case! {
+//     /// successful unlink(2) updates ctime (symlink).
+//     update_ctime_success_symlink
+// }
+// fn update_ctime_success_symlink(ctx: &mut TestContext) {
+//     let path = ctx.create(FileType::Symlink(None)).unwrap();
+//     let link_path = ctx.base_path().join("link");
+//     link(&path, &link_path).unwrap();
+//     assert_ctime_changed(ctx, &link_path, || {
+//         assert!(unlink(&path).is_ok());
+//     });
+// }
+
+crate::test_case! {
+    /// unsuccessful unlink(2) does not update ctime.
+    // unlink/00.t
+    unchanged_ctime_failed, serialized, root => [Regular, Block, Char, Fifo, Socket]
+}
+fn unchanged_ctime_failed(ctx: &mut SerializedTestContext, ft: FileType) {
```

Comment:
It would be possible to write this test without root privileges like this:
* Create directory
* Create file in that directory
* chmod 555 the directory
* assert_ctime_unchanged(|| unlink(dir/file)) # Should fail with EACCES

but since our focus right now is on converting pjdfstest, not improving it, it's ok to defer this for after GSoC

Reply:
That's a very good idea, thanks! I should generify this test anyway, so it wouldn't be a problem to rewrite it.